packages/next/src/server/typescript/rules/metadata.ts

```diff
@@ -4,118 +4,373 @@ import { getSource, getTs, getTypeChecker } from '../utils'
+                file: source,
+                category: ts.DiagnosticCategory.Error,
+                code: NEXT_TS_ERRORS.INVALID_METADATA_EXPORT,
+                messageText: `The Next.js 'metadata' API is not allowed in a client component.`,
+                start: declaration.name.getStart(),
+                length: declaration.name.getWidth(),
+              },
+            ]
+          }
+        }
       }
-    }
-    return []
-  },
+      return []
+    },
+    getSemanticDiagnosticsForExportDeclaration(
+      fileName: string,
+      node: tsModule.ExportDeclaration
+    ): tsModule.Diagnostic[] {
+      const ts = getTs()
+      const source = getSource(fileName)
+      const diagnostics: tsModule.Diagnostic[] = []
 
-  getSemanticDiagnosticsForExportDeclarationInClientEntry(
-    fileName: string,
-    node: tsModule.ExportDeclaration
-  ) {
-    const ts = getTs()
-    const source = getSource(fileName)
-    const diagnostics: tsModule.Diagnostic[] = []
-
-    const exportClause = node.exportClause
-    if (exportClause && ts.isNamedExports(exportClause)) {
-      for (const e of exportClause.elements) {
-        if (['generateMetadata', 'metadata'].includes(e.name.getText())) {
-          diagnostics.push({
-            file: source,
-            category: ts.DiagnosticCategory.Error,
-            code: NEXT_TS_ERRORS.INVALID_METADATA_EXPORT,
-            messageText: `The Next.js '${e.name.getText()}' API is not allowed in a client component.`,
-            start: e.name.getStart(),
-            length: e.name.getWidth(),
-          })
+      const exportClause = node.exportClause
+      if (exportClause && ts.isNamedExports(exportClause)) {
+        for (const e of exportClause.elements) {
+          if (['generateMetadata', 'metadata'].includes(e.name.getText())) {
+            diagnostics.push({
+              file: source,
+              category: ts.DiagnosticCategory.Error,
+              code: NEXT_TS_ERRORS.INVALID_METADATA_EXPORT,
+              messageText: `The Next.js '${e.name.getText()}' API is not allowed in a client component.`,
+              start: e.name.getStart(),
+              length: e.name.getWidth(),
+            })
+          }
         }
       }
-    }
 
-    return diagnostics
+      return diagnostics
+    },
   },
+  server: {
+    getSemanticDiagnosticsForExportVariableStatement(
+      fileName: string,
+      node: tsModule.VariableStatement | tsModule.FunctionDeclaration
+    ): tsModule.Diagnostic[] {
+      const source = getSource(fileName)
+      const ts = getTs()
+
+      if (ts.isFunctionDeclaration(node)) {
+        if (node.name?.getText() === 'generateMetadata') {
+          if (hasCorrectType(node)) {
+            return []
+          }
+
+          const isAsync = node.modifiers?.some(
+            (m) => m.kind === ts.SyntaxKind.AsyncKeyword
+          )
+
+          return [
+            {
+              file: source,
+              category: ts.DiagnosticCategory.Warning,
```

Comment:
Do we have a "recommendation" level or something?

Reply:
I think "Warning" is fine. Especially once we only warn if there's no type annotation. You really shouldn't have any module-level exports that are untyped generally (see TypeScript's isolated declarations effort) and especially not for frameworks that rely on a certain shape. It's just unnecessary pain later on when we want to make breaking changes.